dlt/destinations/snowflake/sql_client.py

```diff
@@ -0,0 +1,164 @@
+from dlt.destinations.snowflake import capabilities
+
+
+class SnowflakeSqlClient(SqlClientBase[snowflake_lib.SnowflakeConnection], DBTransaction):
+
+    dbapi: ClassVar[DBApi] = snowflake_lib
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(self, dataset_name: str, credentials: SnowflakeCredentials) -> None:
+        super().__init__(dataset_name)
+        self._conn: snowflake_lib.SnowflakeConnection = None
+        self.credentials = credentials
+
+    def open_connection(self) -> snowflake_lib.SnowflakeConnection:
+        self._conn = snowflake_lib.connect(
+            user=self.credentials.username,
+            password=self.credentials.password,
+            account=self.credentials.host,
+            warehouse=self.credentials.warehouse,
+            database=self.credentials.database,
+            schema=self.fully_qualified_dataset_name(),
+            role=self.credentials.role
+        )
+        return self._conn
+
+    @raise_open_connection_error
+    def close_connection(self) -> None:
+        if self._conn:
+            self._conn.close()
+            self._conn = None
+
+    @contextmanager
+    def begin_transaction(self) -> Iterator[DBTransaction]:
+        try:
+            self._conn.autocommit(False)
+            yield self
+            self.commit_transaction()
+        except Exception:
+            self.rollback_transaction()
+            raise
+
+    @raise_database_error
+    def commit_transaction(self) -> None:
+        self._conn.commit()
+        self._conn.autocommit(True)
+
+    @raise_database_error
+    def rollback_transaction(self) -> None:
+        self._conn.rollback()
+        self._conn.autocommit(True)
+
+    @property
+    def native_connection(self) -> "snowflake_lib.SnowflakeConnection":
+        return self._conn
+
+    def has_dataset(self) -> bool:
+        query = """
+                SELECT 1
+                    FROM INFORMATION_SCHEMA.SCHEMATA
+                    WHERE schema_name = %s;
+                """
+        rows = self.execute_sql(query, self.fully_qualified_dataset_name(escape=False))
+        return len(rows) > 0
+
+    def create_dataset(self) -> None:
+        self.execute_sql("CREATE SCHEMA %s" % self.fully_qualified_dataset_name())
+
+    def drop_dataset(self) -> None:
+        self.execute_sql("DROP SCHEMA %s CASCADE;" % self.fully_qualified_dataset_name())
+
+    def execute_sql(self, sql: AnyStr, *args: Any, **kwargs: Any) -> Optional[Sequence[Sequence[Any]]]:
+        with self.execute_query(sql, *args, **kwargs) as curr:
+            if curr.description is None:
+                return None
+            else:
+                f = curr.fetchall()
+                return f
+
+    @contextmanager
+    @raise_database_error
+    def execute_query(self, query: AnyStr, *args: Any, **kwargs: Any) -> Iterator[DBApiCursor]:
+        curr: DBApiCursor = None
+        db_args = args if args else kwargs if kwargs else None
```

Comment:
I'd deal with the multiple statements here. Unfortunately snowflake does not really follow dbapi spec regarding multiple results sets in cursors. instead they just returns many cursors. so let's do the following:
1. if dbargs is None then use `execute_string`: https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#object-connection - it returns many cursor so let's return last cursor (the most typical use case)
2. else just execute assuming that all parametrized queries are single statement

this will IMO cover all DDL and merge jobs (they are not parametrized). all other parametrized statements are single. (if there's something in tests then rather fix the tests)

i prefer this than hacking the code to cover for snowflake not implementing dbapi

Reply:
Turns out multi statements are actually supported! Only documented in the rest API, https://docs.snowflake.com/en/developer-guide/sql-api/submitting-multiple-statements

But also works in python: `cur.execute("SELECT 1; SELECT 2;", num_statements=0)`  

dbapi cursor seems to work normally with this, `cursor.nextset()`, etc works.  

`execute_string` splits the string and sends one query at a time so this is better.